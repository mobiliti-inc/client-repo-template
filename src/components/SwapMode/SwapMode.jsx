/* eslint no-underscore-dangle: 0 */
/* eslint-disable camelcase */

import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router';
import moment from 'moment';

import AppFrame from '../../components/AppFrame/AppFrame';
import LoadingSpinner from '../../components/LoadingSpinner/LoadingSpinner';
import Alert from '../../components/Alert/Alert';

import {
	getMarkets,
	refreshAccessToken,
	getUserInformation,
	selectUserLocation,
} from '../../actions';
import handleAuthRequiredApiCall from '../../utils/handleAuthRequiredApiCall';

import { MarketModel } from '../../models';
import styles from './SwapMode.scss';

export const SwapContext = React.createContext({});

const currentYear = moment().year();

class SwapMode extends Component {
	static propTypes = {
		children: PropTypes.node.isRequired,
		swapMode: PropTypes.bool.isRequired,
		markets: PropTypes.arrayOf(MarketModel).isRequired,
		subscriber: PropTypes.shape({
			fetched: PropTypes.bool,
			fetching: PropTypes.bool,
			user: PropTypes.shape({
				first_name: PropTypes.string,
				last_name: PropTypes.string,
				email: PropTypes.email,
				phone: PropTypes.phone,
				dob: PropTypes.string,
			}),
		}).isRequired,
		age: PropTypes.shape({
			fetched: PropTypes.bool,
			data: PropTypes.shape({
				age: PropTypes.string
			})
		}),
		swap: PropTypes.shape({
			swap: PropTypes.bool.isRequired,
			data: PropTypes.objectOf(PropTypes.any).isRequired,
		}).isRequired,
		authData: PropTypes.shape({
			email: PropTypes.string,
			accessToken: PropTypes.string,
			refreshToken: PropTypes.string,
			subscriber_id: PropTypes.number,
		}).isRequired,
		isTokenRefreshed: PropTypes.bool.isRequired,
		history: PropTypes.shape({
			goBack: PropTypes.func.isRequired,
			push: PropTypes.func.isRequired,
			location: PropTypes.shape({
				hash: PropTypes.string,
				key: PropTypes.string,
				pathname: PropTypes.string,
				search: PropTypes.string,
			})
		}).isRequired,
		getUserInformation: PropTypes.func.isRequired,
		selectUserLocation: PropTypes.func.isRequired,
		getMarkets: PropTypes.func.isRequired,
	}

	static defaultProps = {
		age: {
			data: {
				age: '',
			},
			fetched: false,
		},
	}

	state = {
		age: '',
		showSwapModal: true,
	};

	componentDidMount() {
		const { swap } = this.props.swap;
		if (swap) {
			this.fetchMarketsIfNoneFound();
			this.fetchUserInfo();
		} else {
			this.props.history.push('/my-vehicles');
		}
	}

	componentDidUpdate = (prevProps) => {
		const { subscriber: { fetching, fetched, }, isTokenRefreshed } = this.props;
		if (fetched && !fetching && prevProps.fetched !== fetched) {
			this.checkUserAge();
		}

		if (isTokenRefreshed && prevProps.isTokenRefreshed !== isTokenRefreshed) {
			this.fetchUserInfo();
		}
	}

	fetchMarketsIfNoneFound = () => {
		if (this.props.markets.length === 0) {
			this.props.getMarkets();
		}
	}

	fetchUserInfo = () => {
		const {
			authData: {
				refreshToken, accessToken,
			},
			history,
		} = this.props;
		handleAuthRequiredApiCall(
			accessToken,
			refreshToken,
			history,
			this.props.getUserInformation,
			refreshAccessToken,
		);
	}

	checkUserAge = () => {
		const { user: { dob } } = this.props.subscriber;
		const formattedDob = moment(dob, 'MM/DD/YYYY');
		const initialAge = '25+';
		if (!this.state.age) {
			if (dob && formattedDob._isValid) {
				const dobYear = `${currentYear - formattedDob.year()}`;
				this.setState({ age: dobYear });
			} else if (this.props.age.fetched) {
				this.setState({ age: this.props.age.data.age });
			} else {
				this.setState({ age: initialAge });
			}
		}
	}

	render() {
		const {
			children,
			subscriber: { fetching, fetched },
			swapMode,
			swap: { swap },
		} = this.props;

		const { age, showSwapModal } = this.state;
		return (
			<AppFrame {...this.props}>
				{fetched && !fetching && age && swap ?
					(
						<SwapContext.Provider value={{
							userAge: age,
							swapMarket: this.props.swap,
							swapMode,
						}}
						>
							<Alert
								show={showSwapModal}
								title="Swap Mode"
								subTitle="You are currently Swapping your vehicle"
								closable={false}
							/>
							{children}
						</SwapContext.Provider>
					)
					:
					(
						<div styleName="loader-container">
							<LoadingSpinner show blue />
						</div>
					)
				}
			</AppFrame>
		);
	}
}

const mapStateToProps = state => ({
	subscriber: state.subscriber.data,
	authData: state.auth.userData,
	isTokenRefreshed: state.auth.isTokenRefreshed,
	markets: state.markets.data,
	market: state.market,
	swap: state.swap,
});

export default withRouter(connect(
	mapStateToProps,
	{
		getMarkets,
		getUserInformation,
		selectUserLocation,
		refreshAccessToken,
	}
)(CSSModules(SwapMode, styles, { allowMultiple: true })));

export { SwapMode };
