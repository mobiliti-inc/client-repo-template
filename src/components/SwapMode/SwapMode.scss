.loader-container {
    display: flex;
    justify-content: center;
    align-items: center;
    align-content: center;
    height: 70vh;
}