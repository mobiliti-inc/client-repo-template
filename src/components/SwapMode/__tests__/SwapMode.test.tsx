import { SwapMode } from '../SwapMode';

describe('SwapMode', () => {
	const props = {
		children: <div />,
		swapMode: true,
		markets: [{
			id: 1,
			market_name: 'Texas'
		}],
		subscriber: {
			fetched: false,
			fetching: false,
			user: {
				first_name: 'John',
				last_name: 'Doe',
				email: 'john.doe@test.com',
				phone: '123',
				dob: '01/01/1999',
			},
		},
		age: {
			fetched: false,
			data: {
				age: '12'
			}
		},
		swap: {
			swap: true,
			data: {},
		},
		authData: {
			email: 'joe.doe@test.com',
			accessToken: 'token',
			refreshToken: 'token',
			subscriber_id: 1,
		},
		isTokenRefreshed: false,
		history: {
			goBack: jest.fn(),
			push: jest.fn(),
			location: {
				hash: '',
				key: '1',
				pathname: '/',
				search: '/',
			}
		},
		getUserInformation: jest.fn(),
		selectUserLocation: jest.fn(),
		getMarkets: jest.fn(),
	};

	const component = <SwapMode {...props} />;

	it('renders', () => {
		expect(component).toMatchSnapshot();
	});

	it('should show swap modal in case the url matches `/my-vehicles`', () => {
		const swap = {
			swap: false,
			data: {},
		};
		const subject = shallow(<SwapMode {...props} swap={swap} />);
		expect(subject.instance().props.history.push).toBeCalled();
	});

	it('should fetch markets on mount', () => {
		const subject = shallow(<SwapMode {...props} markets={[]} />);
		expect(subject.instance().props.getMarkets).toBeCalled();
	});

	it('should select age based on user dob', () => {
		const subject = shallow(<SwapMode {...props} />);
		subject.setProps({ subscriber: { fetched: true, fetching: false, user: { dob: '01/01/1999' } } });
		expect(subject.state('age')).toEqual('20');
	});

	it('should check if we have age stored in dob is missing', () => {
		const subject = shallow(<SwapMode {...props} />);
		subject.setProps({ subscriber: { fetched: true, fetching: false, user: { dob: '' } }, age: { fetched: true, data: { age: '23' } } });
		expect(subject.state('age')).toEqual('23');
	});

	it('should set initial age to 25+', () => {
		const subject = shallow(<SwapMode {...props} />);
		subject.setProps({ subscriber: { fetched: true, fetching: false, user: { dob: '' } }, age: { fetched: false } });
		expect(subject.state('age')).toEqual('25+');
	});
});
