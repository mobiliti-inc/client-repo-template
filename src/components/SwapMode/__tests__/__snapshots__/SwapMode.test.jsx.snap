// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SwapMode renders 1`] = `
<SwapMode
  age={
    Object {
      "data": Object {
        "age": "12",
      },
      "fetched": false,
    }
  }
  authData={
    Object {
      "accessToken": "token",
      "email": "joe.doe@test.com",
      "refreshToken": "token",
      "subscriber_id": 1,
    }
  }
  getMarkets={[MockFunction]}
  getUserInformation={[MockFunction]}
  history={
    Object {
      "goBack": [MockFunction],
      "location": Object {
        "hash": "",
        "key": "1",
        "pathname": "/",
        "search": "/",
      },
      "push": [MockFunction],
    }
  }
  isTokenRefreshed={false}
  markets={
    Array [
      Object {
        "id": 1,
        "market_name": "Texas",
      },
    ]
  }
  selectUserLocation={[MockFunction]}
  subscriber={
    Object {
      "fetched": false,
      "fetching": false,
      "user": Object {
        "dob": "01/01/1999",
        "email": "john.doe@test.com",
        "first_name": "John",
        "last_name": "Doe",
        "phone": "123",
      },
    }
  }
  swap={
    Object {
      "data": Object {},
      "swap": true,
    }
  }
  swapMode={true}
>
  <div />
</SwapMode>
`;
